First cut at deployment module

hartmans opened the pull request:
This is ready for @fxdgear to take a look at the general direction.
There's still a lot of work to do.
Provide way to find all the objects that should be deployed:

* Deployable is a new protocol for what deployable objects should have as an interface

* DeployableFinder is a plugin mechanism to find all the deployables

* MachineDeployableFinder finds all the machines.

* find_deployables is a top level interface to use the DeployableFinder mechanism.

fxdgear commented on carthage/deployment.py:
```diff
@@ -127,24 +135,73 @@ class MachineDeployableFinder(DeployableFinder):
+    async def find(self, ainjector):
         from .machine import Machine
-        result = await self.ainjector.filter_instantiate_async(Machine, ['host'], stop_at=stop_at, ready=False)
+        result = await ainjector.filter_instantiate_async(Machine, ['host'], stop_at=ainjector, ready=False)
```
There's a bit of magic on this `filter_instantiate_async` I need to get my head around still. 

sometimes the 2nd object (predicate) can be a list and sometimes a callable. 

I'm gonna go read the `find` method a bit more to better understand it..

fxdgear commented on carthage/deployment.py:
```diff
@@ -0,0 +1,494 @@
+        return callback
+
+    def find_callback(self, deployable: Deployable):
+        def callback(future):
```
I noticed the method `callback` inside both `find_callback` and `method_callback` are eerily similar. 

The only differences being how the `else` is handled as well as on line 167 you set `exception=exception.__context__` and on line 189 you set `exception=exception`. 

What is the difference between using `exception.__context__` and just `exception`?

hartmans commented:
>>>>> "Nick" == Nick Lang ***@***.***> writes:
    Nick> I noticed the method `callback` inside both `find_callback`
    Nick> and `method_callback` are eerily similar.

Do you think it's worth trying to make them reuse more code?
If so, how?
I'd probably pass in a success callback into the callback, which for a
lisp person would be entirely natural, but might be stretching Python a
bit.

> The only differences being how the `else` is handled as well
    Nick> as on line 167 you set `exception=exception.__context__` and
    Nick> on line 189 you set `exception=exception`.

> What is the difference between using `exception.__context__`
> and just `exception`?

They both should be exception.__cause__.  We're effectively throwing
away the outer InjectionFailed exception and getting closer to the root
cause.  The interesting thing in the outer exception will be what object
we were working on (the deployable which we capture explicitly) and the
dependency path (which we also capture).  So it makes the failure report
less verbose.  It might even be desirable to actually discard an entire
chain of InjectionFaileds until we get to something else.

fxdgear commented:
> Do you think it's worth trying to make them reuse more code? If so, how? I'd probably pass in a success callback into the callback, which for a lisp person would be entirely natural, but might be stretching Python a bit.

I'd argue for pragmatism at this point and in the future I would probably look for a way to reuse it a bit more if the need arises. 

I was just curious. There were a couple differences which made it seem like this was purposeful as opposed to accidental. :)

hartmans commented:
@fxdgear  I have rebased and cleaned up the patches as well as completed both deployment and destroy.
I think this is ready to merge, and am giving you a chance to go over it before we do that.

fxdgear commented on carthage/deployment.py:
```diff
@@ -0,0 +1,686 @@
+                        logger.debug("%s not deleted: deletion policy retain")
+                        raise IgnoreDeployable
+                    case DeletionPolicy.warn:
+                        logger.warn("%s retained per deletion policy")
```
missing the variable argument on this warn method call.

fxdgear commented on carthage/deployment.py:
```diff
@@ -0,0 +1,686 @@
+            if not dry_run:
+                match policy:
+                    case DeletionPolicy.retain:
+                        logger.debug("%s not deleted: deletion policy retain")
```
missing the variable argument on this warn method call.

fxdgear reviewed (commented):
The test_deployment was really helpful in understanding how all this fits together

found a couple nitpicks

But thanks again for letting me see how this is all going to work!
